CRM/Upgrade/Incremental/MessageTemplates.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 5                                                  |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2018                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2018
+ */
+class CRM_Upgrade_Incremental_MessageTemplates {
+
+  /**
+   * Version we are upgrading to.
+   *
+   * @var string
+   */
+  protected $upgradeVersion;
+
+  /**
+   * @return string
+   */
+  public function getUpgradeVersion() {
+    return $this->upgradeVersion;
+  }
+
+  /**
+   * @param string $upgradeVersion
+   */
+  public function setUpgradeVersion($upgradeVersion) {
+    $this->upgradeVersion = $upgradeVersion;
+  }
+
+  /**
+   * CRM_Upgrade_Incremental_MessageTemplates constructor.
+   *
+   * @param string $upgradeVersion
+   */
+  public function __construct($upgradeVersion) {
+    $this->setUpgradeVersion($upgradeVersion);
+  }
+
+  /**
+   * Get any templates that have been updated.
+   *
+   * @return array
+   */
+  protected function getTemplateUpdates() {
+    return [
+      [
+        'version' => '5.4.alpha1',
+        'template' => 'membership_online_receipt',
+        'upgrade_descriptor' => ts('Use email greeting at top where available'),
+        'type' => 'text',
+      ],
+    ];
+  }
+
```

Comment:
1. Is the idea that this function would be expanded each time there is a template update to add the details here?  If so, could we move this to a separate file or the start or end of this file to make it a bit more obvious that this is intended to be updated by PR authors?

2. Thinking about the situation where there are multiple templates to upgrade with the same message.  Do we need 'type' separately?  If not, we could change 'template' to be `'templates' => ['membership_online_receipt_text', 'membership_online_receipt_html', ...]`. That would need some other changes in the code but would make this config block more concise.

3. Is `[]` now preferred over `array()`?

Reply:
@aydun 
1) yes - I was expecting it to expand over time & I kind of left the determination as to how that would be recorded for the next iteration 
2) hmm - yeah - I'm on the fence - I guess the original function (which was my starting point) used the type because in addition to the obvious (text & html) there is sometimes _subject . I'm not a big fan of string wrangling - not sure on that
3) yes

Reply:
1. ok
2. ok, how would feel about this? :
```
    [
      'version' => '5.4.alpha1',
      'upgrade_descriptor' => ts('Use email greeting at top where available'),
      'templates' => [
         ['name' => 'membership_online_receipt', 'type' => 'text'],
         ['name' => 'membership_online_receipt', 'type' => 'html'],
         ['name' => 'membership_online_receipt', 'type' => 'subject'],
      ],
    ]
```
3. thanks

Reply:
@aydun done!  I've pushed it as an update

(note it specifies an html file update as well as the text one - the intention IS to add one, but also that no harm should be done if it is over-specified)

Reply:
Nice!  Although we need a further change to handle the messages - see new comment.